aria/orchestrator/execution_plugin/ctx_proxy/server.py

```diff
@@ -170,101 +178,59 @@ def _process_ctx_request(ctx, args): # pylint: disable=too-many-branches,too-man
-                callable_args = args[index:index + spec_args_count]
-                # Note: we might actually have fewer args than the args_count, but that could be OK
-                # if the user expects subsequent args to have default values
-
-            args_count = len(callable_args)
-            if args_count > 1:
-                index += args_count - 1
-
-            current = current(*callable_args, **callable_kwargs)
-
-        else:
-            raise RuntimeError('`{0}` cannot be processed in {1}'.format(arg, args))
-
-        index += 1
-
-    if callable(current):
-        current = current()
+    # Parse all arguments
+    while len(args) > 0:
+        obj, args = _parse_argument(obj, args, modifying)
 
     if modifying:
-        if hasattr(current, '__setitem__'):
-            # Modify dict value
-            current[modifying_key] = modifying_value
-        else:
+        if hasattr(obj, '__setitem__'):
+            # Modify item value (dict, list, and similar)
+            if isinstance(obj, (list, tuple)):
+                modifying_key = int(modifying_key)
+            obj[modifying_key] = modifying_value
+        elif hasattr(obj, modifying_key):
             # Modify object attribute
-            setattr(current, modifying_key, modifying_value)
-
-    return current
+            setattr(obj, modifying_key, modifying_value)
+        else:
+            raise ProcessingError('Cannot modify `{0}` of `{1!r}`'
+                                  .format(modifying_key, obj))
 
+    return obj
 
-def _desugar_attr(obj, attr):
-    if not isinstance(attr, basestring):
-        return None
-    if hasattr(obj, attr):
-        return attr
-    attr = attr.replace('-', '_')
-    if hasattr(obj, attr):
-        return attr
-    return None
 
+def _parse_argument(obj, args, modifying):
+    args = list(args)
+    arg = args.pop(0)
 
-def _is_int(arg):
-    try:
-        int(arg)
-    except ValueError:
-        return False
-    return True
+    # Call?
+    if arg == '[':
+        # TODO: should there be a way to escape "[" and "]" in case they are needed as real
+        # arguments?
+        try:
+            closing_index = args.index(']')
+        except ValueError:
+            raise ParsingError('Opening "[" without a closing "]')
+        callable_args = args[:closing_index]
+        args = args[closing_index + 1:]
+        if not callable(obj):
+            raise ProcessingError('Used "[" and "] on an object that is not callable')
+        return obj(*callable_args), args
+
+    # Attribute?
+    if isinstance(arg, basestring):
+        if hasattr(obj, arg):
+            return getattr(obj, arg), args
+        arg_sugared = arg.replace('-', '_')
+        if hasattr(obj, arg_sugared):
+            return getattr(obj, arg_sugared), args
+
+    # Item? (dict, lists, and similar)
+    if hasattr(obj, '__getitem__'):
+        if modifying and hasattr(obj, 'has_key') and (not obj.has_key(arg)):
```

Comment:
Not a good idea. :( Because the "not in" requires an implementation of __iter__, but __iter__ might not necessarily adhere to the dict contract of iterating only keys. (I did add the "right" kind of __iter__ to the instrumented wrapper, but nobody can promise us that other objects would behave as expected here.